### DISCUSSION TITLE
Roslyn analyser for querying

### DISCUSSION THREAD

**Discussion by loraderon** [ravendb/ravendb#17089] [Ideas]
It's not immediately clear to a new user what can be used when querying. Errors are surfaced when the code executes instead of at compile time.

It would be very helpful if there was an official roslyn analyser that could detect invalid queries such as:
```cs
var bars = new[] { 1, 2, 3 };
var foos = await session
  .Query<Foo>()
  .Where(foo => bars.Contains(foo.Bar))
  .ToArrayAsync();
```

And possible provide a code fix:
```cs
var bars = new[] { 1, 2, 3 };
var foos = await session
  .Query<Foo>()
  .Where(foo => foo.Bar.In(bars))
  .ToArrayAsync();
```

**Comment by ayende** [comment]
We would be very happy to accept a pull request for this.
I don't think we currently have anything like that in the road map, and we are fully booked in terms of what we can do for the near future.

**Comment by loraderon** [comment]
This could also help users with other queries that the user thinks will work until it fails during runtime.

> To properly handle your case you have to create a static fanout index
> Currently we don't support auto map fanout indexes (automatically created for queries), but there's an [open issue](https://issues.hibernatingrhinos.com/issue/RavenDB-21773/Handling-of-fanout-auto-map-indexes) for this.

#19548

> Yes, this is the expected behavior. This is because we are flattening out the results of the index and matching on that.

#18208 

> You are operating under the assumption that RavenDB will index into a nested object. That isn't how it works, I'm afraid. If you are passing an object to RavenDB, it will be indexed as a single JSON object.

#18279
